src/Debug.h

```diff
@@ -71,12 +72,47 @@ class Debug
         bool forceAlert; ///< the current debugs() will be a syslog ALERT
     };
 
-    /// whether debugging the given section and the given level produces output
+    /// Preserves 'early' debugs() message and context for future writing into the log file.
+    /// 'early' messages are warning/critical debugs() calls performed before the configuration
+    /// is loaded and the log file is opened.
+    class Message
+    {
+    public:
+        /// the maximum number of messages to accumulate
+        static const int MaxCount = 1000;
+
+        Message(const Context &, const char *);
+
+        int level; ///< the debug level
+        int section; ///< the debug section
+        std::string line; ///< the final message (including timestamp and context) for logging
+    };
+
+    typedef std::vector<Message> Messages;
+
+    /// whether the subsequent Debug::Start() may be called
     static bool Enabled(const int section, const int level)
+    {
+        return LevelAllowed(section, level) || EarlyMessagesAllowed(level);
```

Comment:
However this would not work with '-X', when Debug::Levels[section] is early initialized with '9' but we need still to distinguish <=DBG_IMPORTANT early messages. Also making DBG_IMPORTANT as a default value would cause appearing such messages on startup (but this is probably less important). Anyway, I think we cannot remove these 'early' checks completely within debugs() becase we need to decide whether to store them or not.

Reply:
As we discussed, it is not easy to initialize  Debug::Levels[section] with '-1'.  Can we use std::vector instead and a static method for accessing it? For example:

```
int Debug::Levels(int idx)
{
    static vector<int> DebugLevels(MAX_DEBUG_SECTIONS, 1);
    return DebugLevels[idx];
}
```
There is an overhead due to a function call but it probably can be inlined.

Reply:
Unfortunately, moving the levels container inside a method is essentially the same as (and, in practice, worse than) adding that explicit EarlyMessagesAllowed() (i.e. "have I already initialized the levels array?") check that the PR already has (and I was trying to eliminate). Even if the compiler inlines this Levels() method, it would have to inline that implicit check in nearly every caller -- the compiler cannot know which caller is going to be executed first and, hence, would have to check initialization on nearly every debugs() call.

You can see this check inlined (and duplicated) inside foo() and bar() at https://godbolt.org/z/FpAOw7:

```
foo():
        movzx   eax, BYTE PTR guard variable for Levels()::HiddenLevels[rip]
        test    al, al
        je      .L20
        mov     rax, QWORD PTR Levels()::HiddenLevels[rip]
        mov     eax, DWORD PTR [rax+4]
        ret
.L20:
        push    rbp
        mov     edi, OFFSET FLAT:guard variable for Levels()::HiddenLevels
        call    __cxa_guard_acquire
        ... more initialization code follows ...
```

Compare the above foo() code prefix with trivial foo2() and bar2() that use a globally-initialized array:

```
foo2():
        mov     rax, QWORD PTR ExposedLevels[rip]
        mov     eax, DWORD PTR [rax+4]
        ret
```

foo() and foo2() are essentially examples of debugs() callers, one for each initialization approach.

----

Remove the `-O2` compiler option (or the `inline` keyword for the Levels() method) to see the much smaller non-inlined foo() and bar() -- I suspect the compiler does not inline by default because it costs so much to inline in this case.

Reply:
> Alex: ... change DBG_IMPORTANT to become zero (and DBG_CRITICAL to -1)

For the record, I no longer consider the above idea to be viable. My initial analysis was missing the fact that raw level values are often used in squid.conf and are logged to cache.log. We should keep them stable.